In [1]:
from keras.datasets import mnist
import numpy as np

(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Se obtiene una muestra con igual número de: cincos vs no cincos
# para el conjunto de entrenamiento y el de test
idx_y_train_5 = np.where(y_train == 5)[0]
idx_y_train_n5 = np.where(y_train != 5)[0]
idx_y_train_n5 = np.random.choice(idx_y_train_n5, idx_y_train_5.shape[0], replace=False)

idx_y_train = np.concatenate([idx_y_train_5, idx_y_train_n5])
np.random.shuffle(idx_y_train)

x_train = x_train[idx_y_train]
y_train = y_train[idx_y_train]
print(x_train.shape)

idx_y_test_5 = np.where(y_test == 5)[0]
idx_y_test_n5 = np.where(y_test != 5)[0]
idx_y_test_n5 = np.random.choice(idx_y_test_n5, idx_y_test_5.shape[0], replace=False)

idx_y_test = np.concatenate([idx_y_test_5, idx_y_test_n5])
np.random.shuffle(idx_y_test)

x_test = x_test[idx_y_test]
y_test = y_test[idx_y_test]
print(x_test.shape)

y_train_binary = (y_train == 5).astype(int)
y_test_binary = (y_test == 5).astype(int)

11490434/11490434 [==============================] - 0s 0us/step
(10842, 28, 28)
(1784, 28, 28)


In [11]:
from keras.models import Sequential
from keras.layers import Flatten, Dense
from keras.metrics import F1Score

model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(1)
])

print(model.summary())

model.compile(loss='binary_crossentropy', metrics=['accuracy'])
model.fit(x_train, y_train_binary, epochs = 15, batch_size = 250, validation_split = 0.1)

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_5 (Flatten)         (None, 784)               0         
                                                                 
 dense_5 (Dense)             (None, 1)                 785       
                                                                 
Total params: 785 (3.07 KB)
Trainable params: 785 (3.07 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
None
Epoch 1/20
40/40 [==============================] - 1s 6ms/step - loss: 6.8035 - accuracy: 0.5543 - val_loss: 6.7013 - val_accuracy: 0.5622
Epoch 2/20
40/40 [==============================] - 0s 3ms/step - loss: 6.0436 - accuracy: 0.6048 - val_loss: 5.7297 - val_accuracy: 0.6267
Epoch 3/20
40/40 [==============================] - 0s 3ms/step - loss: 5.0930 - accuracy: 0.6658 - val_loss: 4.8405 - val_accuracy: 0.

In [12]:
from sklearn.metrics import confusion_matrix, \
precision_score, recall_score,f1_score

y_pred = (model.predict(x_test) >= 0.5).astype(int)[:,0]

conf_matrix = confusion_matrix(y_test_binary, y_pred)
print("Matriz de confusión:")
print(conf_matrix)

precision = precision_score(y_test_binary, y_pred)
print("Precision_1:", precision)
recall = recall_score(y_test_binary, y_pred)
print("Recall_1:", recall)
f1_s = f1_score(y_test_binary, y_pred)
print("F1-score_1:", f1_s)


precision_0 = precision_score(y_test_binary, y_pred, pos_label=0)
print("Precision_0:", precision_0)
recall_0 = recall_score(y_test_binary, y_pred, pos_label=0)
print("Recall_0:", recall_0)
f1_s_0 = f1_score(y_test_binary, y_pred, pos_label=0)
print("F1-score_0:", f1_s_0)

56/56 [==============================] - 0s 1ms/step
Matriz de confusión:
[[798  94]
 [136 756]]
Precision_1: 0.8894117647058823
Recall_1: 0.8475336322869955
F1-score_1: 0.8679678530424799
Precision_0: 0.854389721627409
Recall_0: 0.8946188340807175
F1-score_0: 0.8740416210295728


### Reporte de entrenamientos

Entrenamiento 1:
epochs = 4, batch size = 2000, validation split = 0.9


val_accuracy: 0.5275

Precision_1: 0.5208078335373317

Recall_1: 0.9540358744394619

F1-score_1: 0.6737925574030088

accuracy: 0.5424

Precision_0: 0.5197066911090742

Recall_0: 0.6356502242152466

F1-score_0: 0.5718608169440241

Entrenamiento 2:
epochs = 20, batch size = 2000, validation split = 0.9

val_accuracy: 0.6634

Precision_1: 0.56118754525706

Recall_1: 0.8688340807174888

F1-score_1: 0.6819181698196216

accuracy: 0.6605

Precision_0: 0.5197066911090742

Recall_0: 0.6356502242152466

F1-score_0: 0.5718608169440241

Entrenamiento 3:
epochs = 4, batch size = 2, validation split = 0.9

val_accuracy: 0.5101

Precision_1: 0.5632911392405063

Recall_1: 0.19955156950672645

F1-score_1: 0.29470198675496684

accuracy: 0.5138

Precision_0: 0.5136239782016349

Recall_0: 0.8452914798206278

F1-score_0: 0.6389830508474575

Entrenamiento 4:
epochs = 4, batch size = 2000, validation split = 0.01

val_accuracy: 0.6147

precision_1: 0.7042577675489068

Recall_1: 0.6860986547085202

F1-score_1: 0.6950596252129472

accuracy 0.62902

Precision_0: 0.6939890710382514

Recall_0: 0.7118834080717489

F1-score_0: 0.7028223574986165



Entrenamiento 5

Buscar un entrenamiento de estos paramétros que optimize las métricas.

1. ephocs = 10, batch_size = 500, validation_split = 0.2 porque es el estándar.
2. El accuracy fue de 0.688 y val_accuracy de 0.6782. Modificamos el batch size para buscar optimizar las métricas. ephocs = 10, batch_size = 250, validation_split = 0.2.
Se obtuvieron los siguientes resultados accuracy: 0.7753, val_accuracy: 0.7381, Precision_1: 0.8345642540620384,
Recall_1: 0.6334080717488789,
F1-score_1: 0.7202039515615041,
Precision_0: 0.7046070460704607,
Recall_0: 0.874439461883408,
F1-score_0: 0.7803901950975488
3. Se usa un conjunto de entrenamiento mayor para que el modelo pueda entrenar con más ejemplos. ephocs = 10, batch_size = 250, validation_split = 0.1. Se obtuvieron los siguientes resultados accuracy: 0.8230, val_accuracy: 0.7806, Precision_1: 0.733920704845815,
Recall_1: 0.9338565022421524,
F1-score_1: 0.8219042920572274,
Precision_0: 0.9090909090909091,
Recall_0: 0.6614349775784754,
F1-score_0: 0.7657365347177157
4. Se aumenta el número de epócas para intentar mejorar el entrenamiento.
ephocs = 15, batch_size = 250, validation_split = 0.1. Se obtuvieron los siguientes resultados accuracy: 0.8555, val_accuracy: 0.8507, Precision_1: 0.733920704845815,
Recall_1: 0.9338565022421524,
F1-score_1: 0.8219042920572274,
Precision_0: 0.9090909090909091,
Recall_0: 0.6614349775784754,
F1-score_0: 0.7657365347177157
5. Se aumenta el número de epocas. ephocs = 20, batch_size = 250, validation_split = 0.1. Se obtuvieron los siguientes resultados accuracy: 0.8548 val_accuracy: 0.8645, Precision_1: 0.8894117647058823,
Recall_1: 0.8475336322869955,
F1-score_1: 0.8679678530424799,
Precision_0: 0.854389721627409,
Recall_0: 0.8946188340807175,
F1-score_0: 0.8740416210295728,
